Better actionable error message on InvalidImageReferenceException: Invalid image reference

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

**Description of the issue**:
An uncheckedException, InvalidImageReferenceException, when pushing to an gcr

**Expected behavior**:
Should be able to successfully push new image to remote gcr

**Steps to reproduce**: <!-- Please provide a minimal and precise series of steps -->
* add the jib gradle plugin
* configure the jib plugin to push image to gcr
* run a gradle jib

**Environment**: <!-- OS, Maven, and other relevant environment information -->
Gradle, MacOS


**`jib-gradle-plugin` Configuration**: <!-- Delete this section if not used -->
```groovy
jib{
    from{
        image = 'openjdk:11-jre-slim'
    }
    to{
        image = "gcr.io/****/$rootProject.name"
    }
    container{
        jvmFlags = ['-Xdebug']
        ports = ['8181']
    }
}
```

```
Caused by: org.gradle.internal.UncheckedException: com.google.cloud.tools.jib.image.InvalidImageReferenceException: Invalid image reference: gcr.io/***/my_app_name
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:67)
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:41)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:106)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:48)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:41)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:705)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:672)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$4.run(ExecuteActionsTaskExecuter.java:338)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:327)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:312)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:75)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:15
```

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/image/InvalidImageReferenceException.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/image/InvalidImageReferenceException.java
@@ -20,6 +20,6 @@ package com.google.cloud.tools.jib.image;
 public class InvalidImageReferenceException extends Exception {
 
   public InvalidImageReferenceException(String reference) {
-    super("Invalid image reference: " + reference);
+    super(reference);
   }
 }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/image/InvalidImageReferenceException.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/image/InvalidImageReferenceException.java
@@ -19,7 +19,14 @@ package com.google.cloud.tools.jib.image;
 /** Thrown when attempting to parse an invalid image reference. */
 public class InvalidImageReferenceException extends Exception {
 
+  private final String reference;
+
   public InvalidImageReferenceException(String reference) {
-    super(reference);
+    super("Invalid image reference: " + reference);
+    this.reference = reference;
+  }
+
+  public String getInvalidReference() {
+    return reference;
   }
 }

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java
@@ -171,7 +171,8 @@ public class BuildDockerTask extends DefaultTask implements JibTask {
           ex);
 
     } catch (InvalidImageReferenceException ex) {
-      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+      throw new GradleException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
     }
   }
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java
@@ -87,8 +87,8 @@ public class BuildDockerTask extends DefaultTask implements JibTask {
 
   @TaskAction
   public void buildDocker()
-      throws InvalidImageReferenceException, IOException, BuildStepsExecutionException,
-          CacheDirectoryCreationException, MainClassInferenceException {
+      throws IOException, BuildStepsExecutionException, CacheDirectoryCreationException,
+          MainClassInferenceException {
     Path dockerExecutable = dockerClientParameters.getExecutablePath();
     boolean isDockerInstalled =
         dockerExecutable == null
@@ -169,6 +169,9 @@ public class BuildDockerTask extends DefaultTask implements JibTask {
           HelpfulSuggestions.forIncompatibleBaseImageJavaVesionForGradle(
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
+
+    } catch (InvalidImageReferenceException ex) {
+      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
     }
   }
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java
@@ -74,8 +74,8 @@ public class BuildImageTask extends DefaultTask implements JibTask {
 
   @TaskAction
   public void buildImage()
-      throws InvalidImageReferenceException, IOException, BuildStepsExecutionException,
-          CacheDirectoryCreationException, MainClassInferenceException {
+      throws IOException, BuildStepsExecutionException, CacheDirectoryCreationException,
+          MainClassInferenceException {
     // Asserts required @Input parameters are not null.
     Preconditions.checkNotNull(jibExtension);
     TaskCommon.disableHttpLogging();
@@ -148,6 +148,9 @@ public class BuildImageTask extends DefaultTask implements JibTask {
           HelpfulSuggestions.forIncompatibleBaseImageJavaVesionForGradle(
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
+
+    } catch (InvalidImageReferenceException ex) {
+      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
     }
   }
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java
@@ -150,7 +150,8 @@ public class BuildImageTask extends DefaultTask implements JibTask {
           ex);
 
     } catch (InvalidImageReferenceException ex) {
-      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+      throw new GradleException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
     }
   }
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildTarTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildTarTask.java
@@ -167,7 +167,8 @@ public class BuildTarTask extends DefaultTask implements JibTask {
           ex);
 
     } catch (InvalidImageReferenceException ex) {
-      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+      throw new GradleException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
     }
   }
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildTarTask.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildTarTask.java
@@ -96,8 +96,8 @@ public class BuildTarTask extends DefaultTask implements JibTask {
 
   @TaskAction
   public void buildTar()
-      throws InvalidImageReferenceException, BuildStepsExecutionException, IOException,
-          CacheDirectoryCreationException, MainClassInferenceException {
+      throws BuildStepsExecutionException, IOException, CacheDirectoryCreationException,
+          MainClassInferenceException {
     // Asserts required @Input parameters are not null.
     Preconditions.checkNotNull(jibExtension);
     TaskCommon.disableHttpLogging();
@@ -165,6 +165,9 @@ public class BuildTarTask extends DefaultTask implements JibTask {
           HelpfulSuggestions.forIncompatibleBaseImageJavaVesionForGradle(
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
+
+    } catch (InvalidImageReferenceException ex) {
+      throw new GradleException(HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
     }
   }
 

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
@@ -165,10 +165,11 @@ public class BuildDockerMojo extends JibPluginConfiguration {
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
 
-    } catch (InvalidImageReferenceException
-        | IOException
-        | CacheDirectoryCreationException
-        | MainClassInferenceException ex) {
+    } catch (InvalidImageReferenceException ex) {
+      throw new MojoExecutionException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+
+    } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);
 
     } catch (BuildStepsExecutionException ex) {

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
@@ -167,7 +167,7 @@ public class BuildDockerMojo extends JibPluginConfiguration {
 
     } catch (InvalidImageReferenceException ex) {
       throw new MojoExecutionException(
-          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
 
     } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
@@ -157,10 +157,11 @@ public class BuildImageMojo extends JibPluginConfiguration {
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
 
-    } catch (InvalidImageReferenceException
-        | IOException
-        | CacheDirectoryCreationException
-        | MainClassInferenceException ex) {
+    } catch (InvalidImageReferenceException ex) {
+      throw new MojoExecutionException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+
+    } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);
 
     } catch (BuildStepsExecutionException ex) {

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
@@ -159,7 +159,7 @@ public class BuildImageMojo extends JibPluginConfiguration {
 
     } catch (InvalidImageReferenceException ex) {
       throw new MojoExecutionException(
-          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
 
     } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
@@ -135,10 +135,11 @@ public class BuildTarMojo extends JibPluginConfiguration {
               ex.getBaseImageMajorJavaVersion(), ex.getProjectMajorJavaVersion()),
           ex);
 
-    } catch (InvalidImageReferenceException
-        | IOException
-        | CacheDirectoryCreationException
-        | MainClassInferenceException ex) {
+    } catch (InvalidImageReferenceException ex) {
+      throw new MojoExecutionException(
+          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+
+    } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);
 
     } catch (BuildStepsExecutionException ex) {

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
@@ -137,7 +137,7 @@ public class BuildTarMojo extends JibPluginConfiguration {
 
     } catch (InvalidImageReferenceException ex) {
       throw new MojoExecutionException(
-          HelpfulSuggestions.forInvalidImageReference(ex.getMessage()), ex);
+          HelpfulSuggestions.forInvalidImageReference(ex.getInvalidReference()), ex);
 
     } catch (IOException | CacheDirectoryCreationException | MainClassInferenceException ex) {
       throw new MojoExecutionException(ex.getMessage(), ex);

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
@@ -79,7 +79,7 @@ public class HelpfulSuggestions {
     return suggest(
         "Invalid image reference " + reference,
         "check that the reference is formatted correctly according to "
-            + "https://docs.docker.com/engine/reference/commandline/tag/#extended-description "
+            + "https://docs.docker.com/engine/reference/commandline/tag/#extended-description\n"
             + "For example, slash-separated name components cannot have uppercase letters");
   }
 

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
@@ -75,6 +75,14 @@ public class HelpfulSuggestions {
             + " or below");
   }
 
+  public static String forInvalidImageReference(String reference) {
+    return suggest(
+        "Invalid image reference " + reference,
+        "check that the reference is formatted correctly according to "
+            + "https://docs.docker.com/engine/reference/commandline/tag/#extended-description "
+            + "For example, slash-separated name components cannot have uppercase letters");
+  }
+
   /**
    * @param messagePrefix the initial message text
    * @param suggestion a suggested fix for the problem described by {@link #messagePrefix}

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
@@ -84,7 +84,7 @@ public class HelpfulSuggestionsTest {
         HelpfulSuggestions.forIncompatibleBaseImageJavaVesionForMaven(8, 11));
     Assert.assertEquals(
         "Invalid image reference gcr.io/invalid_REF, perhaps you should check that the reference "
-            + "is formatted correctly according to https://docs.docker.com/engine/reference/commandline/tag/#extended-description "
+            + "is formatted correctly according to https://docs.docker.com/engine/reference/commandline/tag/#extended-description\n"
             + "For example, slash-separated name components cannot have uppercase letters",
         HelpfulSuggestions.forInvalidImageReference("gcr.io/invalid_REF"));
     Assert.assertEquals("messagePrefix", TEST_HELPFUL_SUGGESTIONS.none());

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
@@ -82,6 +82,11 @@ public class HelpfulSuggestionsTest {
             + "parameter, or set maven-compiler-plugin's '<target>' or '<release>' version "
             + "to 8 or below in your build configuration",
         HelpfulSuggestions.forIncompatibleBaseImageJavaVesionForMaven(8, 11));
+    Assert.assertEquals(
+        "Invalid image reference gcr.io/invalid_REF, perhaps you should check that the reference "
+            + "is formatted correctly according to https://docs.docker.com/engine/reference/commandline/tag/#extended-description "
+            + "For example, slash-separated name components cannot have uppercase letters",
+        HelpfulSuggestions.forInvalidImageReference("gcr.io/invalid_REF"));
     Assert.assertEquals("messagePrefix", TEST_HELPFUL_SUGGESTIONS.none());
     Assert.assertEquals(
         "messagePrefix, perhaps you should use a registry that supports HTTPS so credentials can be sent safely, or set the 'sendCredentialsOverHttp' system property to true",

